engine/crates/engine-v2/schema/src/builder/ids.rs

```diff
@@ -0,0 +1,196 @@
+//! This module is responsible for mapping identifiers between federated schema and the engine v2
+//! schema when the mapping is not 1:1.
+//!
+//! As of 3b56f12c95d334ce6cb46f4f8654ce531a69f975, this only happens when @inaccessible items
+//! are removed.
+
+use crate::ids::*;
+
+pub(super) trait MappableId: Copy {
```

Comment:
IMHO this could probably be simpler with a few changes:
- You could avoid `get_idx` by implementing `usize: From<OriginId>` for the federated ids
- Instead of having an array of existing ids, I would use `Vec<Option<TargetId>` with the same size as the number of `OriginId`. So `fields_mapping = vec![None; federated_graph.fields.len()]` Since the ids are `NonZeroU32` in the schema it's free. It's simpler and smaller than a `[usize]`. You only need custom logic for the ranges.

With this you could just implement:
```
impl Index<OriginId> for IdMapper { type Target = Option<TargetId> }

impl Index<(OriginId, usize)> for IdMapper { type Target = IdRange<TargetId> }
```

and have a nicer API to work with later

Reply:
I don't think you would need a trait with this either, just a macro implementing it for all id pairs like almost everything related to ids would work imho

Reply:
The idea behind this representation is that you have Vec's that will grow to `len() == inaccessible_count` instead of the same size as the vec of elements. But yeah, I'll try again (I tried this but the branch got messy with other things so ended up reverted), it's worth trying to simplify.

Reply:
ah, didn't understood it

Reply:
Ah smart. What's unclear is `get_map` and `get_map_mut` then. The variables names `skipped` felt weird indeed, make a lot more sense now.
I think you could still avoid the trait, but up to you.

I would rename `map` to `skipped_ids`, would be a lot clearer IMHO

Reply:
done